// node-fzf/src/main.js
// used to read keyboard input while at the same time
// reading piped stdin input and printing to stdout
const keypress = require( 'keypress' )
const ttys = require( 'ttys' )

const stdin = ttys.stdin
const stdout = ttys.stdout

// get printed width of text
// ex. 漢字 are 4 characters wide but still
// only 2 characters in length
const stringWidth = require( 'string-width' )

// print/render to the terminal
const clc = require( 'cli-color' )

// available filtering modes ( fuzzy by default )
const modes = [ 'fuzzy', 'normal' ]

module.exports = queryUser

// helper to only get user input
module.exports.getInput = getInput

function getInput ( label, callback )
{
  const opts = {
    label: label,
    list: [],
    nolist: true // don't print list/matches
  }

  return queryUser( opts, callback )
}

function queryUser ( opts, callback )
{
  /* opts should reference same object at all times
   * as it will be returned as an api as well that the
   * user can use.
   *
   * a few functions will be added to opts that will
   * work as an API to the user to ex. update the list
   * at a later time.
   *
   * we do it like this instead of return a separate api
   * object in order to support promises when a callback
   * fn is omitted.
   */
  const _opts = opts

  if ( Array.isArray( _opts ) ) {
    _opts.list = _opts
    _opts.mode = 'fuzzy'
  }

  if ( typeof _opts !== 'object' ) {
    // in JavaScript arrays are also a typeof 'object'
    throw new TypeError( 'arg0 has to be an array or an object' )
  }

  _opts.list = _opts.list || []
  _opts.mode = _opts.mode || 'fuzzy'

  const promise = new Promise( function ( resolve, reject ) {
    let originalList = _opts.list || []
    let _list = prepareList( originalList )

    let _input = ''

    // user defined vertical scrolling
    let scrollOffset = 0

    _opts.update = function ( list ) {
      originalList = list
      _opts.list = originalList
      _list = prepareList( originalList )
      render()
    }

    _opts.stop = function () {
      finish()
    }

    // prepare provided list for internal searching/sorting
    function prepareList ( newList ) {
      const list = newList.map( function ( value, index ) {
        return {
          originalValue: value, // text
          originalIndex: index
        }
      } )
      return list
    }

    function finish ( result ) {
      if ( finish.done ) return
      finish.done = true

      stdout.removeListener( 'resize', handleResize )

      stdin.removeListener( 'keypress', handleKeypress )

      stdin.setRawMode && stdin.setRawMode( false )
      stdin.pause()

      if ( !result ) {
        // quit, exit, cancel, abort
        buffer = undefined

        result = {
          selected: undefined,

          // common alternatives for the same thing
          query: buffer,
          search: buffer,
          input: buffer
        }
      }

      if ( callback ) {
        callback( result )
      }

      resolve( result )
    }

    // make `process.stdin` begin emitting "keypress" events
    keypress( stdin )

    // selected index relative to currently matched results
    // (filtered subset of _list)
    let selectedIndex = 0

    // input buffer
    let buffer = ''

    // input cursor position ( only horizontal )
    // relative to input buffer
    let cursorPosition = 0

    // number of items printed on screen, usually ~7
    let _printedMatches = 0

    let _matches = []
    let _selectedItem

    const MIN_HEIGHT = 6

    function getMaxWidth () {
      const mx = stdout.columns - 7
      return Math.max( 0, mx )
    }

    stdout.on( 'resize', handleResize )

    function handleResize () {
      clearTimeout( handleResize.timeout )
      handleResize.timeout = setTimeout( function () {
        cleanDirtyScreen()
        render()
      }, 1 )
    }

    const debug = false

    function handleKeypress ( chunk, key ) {
      debug && console.log( 'chunk: ' + chunk )

      key = key || { name: '' }

      const name = String( key.name )

      debug && console.log( 'got "keypress"', key )

      if ( key && key.ctrl && name === 'c' ) {
        cleanDirtyScreen()
        return finish()
      }

      if ( key && key.ctrl && name === 'z' ) {
        cleanDirtyScreen()
        return finish()
      }

      if ( key && key.ctrl && name === 'l' ) {
        // return stdout.write( clc.reset )
      }

      const view_height = _printedMatches || 10

      if ( key.ctrl ) {
        switch ( name ) {
          case 'h': // backspace
            // ignore
            break

          case 'b': // jump back 1 word
            {
              const slice = buffer.slice( 0, cursorPosition )
              const m = slice.match( /\S+\s*$/ ) // last word
              if ( m && m.index > 0 ) {
                // console.log( m.index )
                cursorPosition = m.index
              } else {
                cursorPosition = 0
              }
            }
            return render()
            break

          case 'j': // down
          case 'n': // down
            selectedIndex += 1
            return render()
            break
          case 'k': // up
          case 'p': // up
            selectedIndex -= 1
            return render()
            break

          case 'l': // right
            // ignore
            break

          case 's':
            // TODO ctrl-s support? switch between match modes?
            {
              // cleanDirtyScreen()
              let i = modes.indexOf( _opts.mode )
              _opts.mode = modes[ ++i % modes.length ]
            }
            return render()
            break

          case 'f': // jump forward 1 word
            {
              const slice = buffer.slice( cursorPosition )
              const m = slice.match( /^\S+\s*/ ) // first word
              if ( m && m.index >= 0 && m[ 0 ] && m[ 0 ].length >= 0 ) {
                // console.log( m.index )
                cursorPosition += ( m.index + m[ 0 ].length )
              } else {
                cursorPosition = buffer.length
              }
            }
            return render()
            break

          case 'd': // down
            // basically intended as page-down
            selectedIndex += view_height
            return render()
            break

          case 'u': // up
            // basically intended as page-up
            selectedIndex -= view_height
            return render()
            break

          case 'a': // beginning of line
            cursorPosition = 0
            return render()
            break

          case 'e': // end of line
            cursorPosition = buffer.length
            return render()
            break

          case 'w': // clear word
            {
              const a = buffer.slice( 0, cursorPosition )
              const b = buffer.slice( cursorPosition )
              const m = a.match( /\S+\s*$/ ) // last word
              if ( m && m.index > 0 ) {
                // console.log( m.index )
                cursorPosition = m.index
                buffer = a.slice( 0, cursorPosition ).concat( b )
              } else {
                cursorPosition = 0
                buffer = b
              }
            }
            return render()
            break

          case 'q': // quit
            cleanDirtyScreen()
            return finish()
        }
      }

      // usually ALT key
      if ( key.meta ) {
        switch ( name ) {
          case 'n': // left arrow key
            scrollOffset--
            return render()

          case 'p': // right arrow key
            scrollOffset++
            return render()
        }
      }

      if ( key.ctrl ) return
      if ( key.meta ) return

      switch ( name ) {
        case 'backspace': // ctrl-h
          {
            const a = buffer.slice( 0, cursorPosition - 1 )
            const b = buffer.slice( cursorPosition )
            buffer = a.concat( b )

            cursorPosition--
            if ( cursorPosition < 0 ) {
              cursorPosition = 0
            }
          }
          return render()
          break

        case 'left': // left arrow key
          cursorPosition--
          if ( cursorPosition < 0 ) cursorPosition = 0
          return render()
          break

        case 'right': // right arrow key
          cursorPosition++
          if ( cursorPosition > buffer.length ) {
            cursorPosition = buffer.length
          }
          return render()
          break

        // text terminals treat ctrl-j as newline ( enter )
        // ref: https://ss64.com/bash/syntax-keyboard.html
        case 'down': // ctrl-j
        case 'enter':
          selectedIndex += 1
          return render()

        case 'up':
          selectedIndex -= 1
          return render()

        case 'esc':
        case 'escape':
          cleanDirtyScreen()
          return finish()

        // hit return key ( aka enter key ) ( aka ctrl-m )
        case 'return': // ctrl-m
          cleanDirtyScreen()

          function transformResult ( match ) {
            return {
              value: match.originalValue,
              index: match.originalIndex
            }
          }

          const result = {
            selected: _selectedItem && transformResult( _selectedItem ) || undefined,

            // common alternatives for the same thing
            query: buffer,
            search: buffer,
            input: buffer
          }

          return finish( result )
      }

      if ( chunk && chunk.length === 1 ) {
        let c = ''
        if ( key.shift ) {
          c = chunk.toUpperCase()
        } else {
          c = chunk
        }

        if ( c ) {
          const a = buffer.slice( 0, cursorPosition )
          const b = buffer.slice( cursorPosition )
          buffer = a.concat( c, b )

          cursorPosition++
          if ( cursorPosition > buffer.length ) {
            cursorPosition = buffer.length
          }
        }

        render()
      }
    }

    stdin.setEncoding( 'utf8' )
    stdin.on( 'keypress', handleKeypress )

    const clcBgGray = clc.bgXterm( 236 )
    const clcFgArrow = clc.xterm( 198 )
    const clcFgBufferArrow = clc.xterm( 110 )
    const clcFgGreen = clc.xterm( 143 )
    // const clcFgMatchGreen = clc.xterm( 151 )
    const clcFgModeStatus = clc.xterm( 110 )
    const clcFgMatchGreen = clc.xterm( 107 )

    // get matches based on the search mode
    function getMatches ( mode, filter, text )
    {
      switch ( mode.trim().toLowerCase() ) {
        case 'normal':
          return textMatches( filter, text )

        case 'fuzzy':
        default:
          // default to fuzzy matching
          return fuzzyMatches( filter, text )
      }
    }

    // get matched list based on the search mode
    function getList ( mode, filter, list )
    {
      // default to fuzzy matching
      switch ( mode.trim().toLowerCase() ) {
        case 'normal':
          return textList( filter, list )

        case 'fuzzy':
        default:
          // default to fuzzy matching
          return fuzzyList( filter, list )
      }
    }

    function fuzzyMatches ( fuzz, text )
    {
      fuzz = fuzz.toLowerCase()
      text = text.toLowerCase()

      let tp = 0 // text position/pointer
      let matches = []

      // nothing to match with
      if ( !fuzz ) return matches

      for ( let i = 0; i < fuzz.length; i++ ) {
        const f = fuzz[ i ]

        for ( ; tp < text.length; tp++ ) {
          const t = text[ tp ]
          if ( f === t ) {
            matches.push( tp )
            tp++
            break
          }
        }
      }

      return matches
    }

    function fuzzyList ( fuzz, list )
    {
      const results = []

      for ( let i = 0; i < list.length; i++ ) {
        const item = list[ i ]

        const originalIndex = item.originalIndex
        const originalValue = item.originalValue

        // get rid of unnecessary whitespace that only takes of
        // valuable scren space
        const normalizedItem = originalValue.split( /\s+/ ).join( ' ' )

        /* matches is an array of indexes on the normalizedItem string
         * that have matched the fuzz
         */
        const matches = fuzzyMatches( fuzz, normalizedItem )

        if ( matches.length === fuzz.length ) {
          /* When the matches.length is exacly the same as fuzz.length
           * it means we have a fuzzy match -> all characters in
           * the fuzz string have been found on the normalizedItem string.
           * The matches array holds each string index position
           * of those matches on the normalizedItem string.
           * ex. fuzz = 'foo', normalizedItem = 'far out dog', matches = [0,4,9]
           */

          let t = normalizedItem

          resu
