// cacheable/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Cacheable: () => Cacheable,
  CacheableEvents: () => CacheableEvents,
  CacheableHooks: () => CacheableHooks,
  CacheableMemory: () => import_memory2.CacheableMemory,
  CacheableStats: () => import_utils2.Stats,
  CacheableSync: () => CacheableSync,
  CacheableSyncEvents: () => CacheableSyncEvents,
  HashAlgorithm: () => import_utils2.HashAlgorithm,
  Keyv: () => import_keyv2.Keyv,
  KeyvCacheableMemory: () => import_memory2.KeyvCacheableMemory,
  KeyvHooks: () => import_keyv2.KeyvHooks,
  calculateTtlFromExpiration: () => import_utils2.calculateTtlFromExpiration,
  createKeyv: () => import_memory2.createKeyv,
  getCascadingTtl: () => import_utils2.getCascadingTtl,
  getOrSet: () => import_utils2.getOrSet,
  hash: () => import_utils2.hash,
  shorthandToMilliseconds: () => import_utils2.shorthandToMilliseconds,
  shorthandToTime: () => import_utils2.shorthandToTime,
  wrap: () => import_utils2.wrap,
  wrapSync: () => import_utils2.wrapSync
});
module.exports = __toCommonJS(index_exports);
var import_memory = require("@cacheable/memory");
var import_utils = require("@cacheable/utils");
var import_hookified2 = require("hookified");
var import_keyv = require("keyv");

// src/enums.ts
var CacheableHooks = /* @__PURE__ */ ((CacheableHooks2) => {
  CacheableHooks2["BEFORE_SET"] = "BEFORE_SET";
  CacheableHooks2["AFTER_SET"] = "AFTER_SET";
  CacheableHooks2["BEFORE_SET_MANY"] = "BEFORE_SET_MANY";
  CacheableHooks2["AFTER_SET_MANY"] = "AFTER_SET_MANY";
  CacheableHooks2["BEFORE_GET"] = "BEFORE_GET";
  CacheableHooks2["AFTER_GET"] = "AFTER_GET";
  CacheableHooks2["BEFORE_GET_MANY"] = "BEFORE_GET_MANY";
  CacheableHooks2["AFTER_GET_MANY"] = "AFTER_GET_MANY";
  CacheableHooks2["BEFORE_SECONDARY_SETS_PRIMARY"] = "BEFORE_SECONDARY_SETS_PRIMARY";
  return CacheableHooks2;
})(CacheableHooks || {});
var CacheableEvents = /* @__PURE__ */ ((CacheableEvents2) => {
  CacheableEvents2["ERROR"] = "error";
  CacheableEvents2["CACHE_HIT"] = "cache:hit";
  CacheableEvents2["CACHE_MISS"] = "cache:miss";
  return CacheableEvents2;
})(CacheableEvents || {});

// src/sync.ts
var import_hookified = require("hookified");
var import_qified = require("qified");
var CacheableSyncEvents = /* @__PURE__ */ ((CacheableSyncEvents2) => {
  CacheableSyncEvents2["ERROR"] = "error";
  CacheableSyncEvents2["SET"] = "cache:set";
  CacheableSyncEvents2["DELETE"] = "cache:delete";
  return CacheableSyncEvents2;
})(CacheableSyncEvents || {});
var CacheableSync = class extends import_hookified.Hookified {
  _qified = new import_qified.Qified();
  _namespace;
  _storage;
  _cacheId;
  /**
   * Creates an instance of CacheableSync
   * @param options - Configuration options for CacheableSync
   */
  constructor(options) {
    super(options);
    this._namespace = options.namespace;
    this._qified = this.createQified(options.qified);
  }
  /**
   * Gets the Qified instance used for synchronization
   * @returns The Qified instance
   */
  get qified() {
    return this._qified;
  }
  /**
   * Sets the Qified instance used for synchronization
   * @param value - Either an existing Qified instance or MessageProvider(s)
   */
  set qified(value) {
    this._qified = this.createQified(value);
  }
  /**
   * Gets the namespace for sync events
   * @returns The namespace or undefined if not set
   */
  get namespace() {
    return this._namespace;
  }
  /**
   * Sets the namespace for sync events and resubscribes if needed
   * @param namespace - The namespace string or function
   */
  set namespace(namespace) {
    if (this._storage && this._cacheId) {
      const oldSetEvent = this.getPrefixedEvent("cache:set" /* SET */);
      const oldDeleteEvent = this.getPrefixedEvent("cache:delete" /* DELETE */);
      void this._qified.unsubscribe(oldSetEvent);
      void this._qified.unsubscribe(oldDeleteEvent);
    }
    this._namespace = namespace;
    if (this._storage && this._cacheId) {
      this.subscribe(this._storage, this._cacheId);
    }
  }
  /**
   * Publishes a cache event to all the cache instances
   * @param data - The cache item data containing cacheId, key, value, and optional ttl
   */
  async publish(event, data) {
    const eventName = this.getPrefixedEvent(event);
    await this._qified.publish(eventName, {
      id: crypto.randomUUID(),
      data
    });
  }
  /**
   * Subscribes to sync events and updates the provided storage
   * @param storage - The Keyv storage instance to update
   * @param cacheId - The cache ID to identify this instance
   */
  subscribe(storage, cacheId) {
    this._storage = storage;
    this._cacheId = cacheId;
    const setEvent = this.getPrefixedEvent("cache:set" /* SET */);
    const deleteEvent = this.getPrefixedEvent("cache:delete" /* DELETE */);
    this._qified.subscribe(setEvent, {
      handler: async (message) => {
        const data = message.data;
        if (data.cacheId !== cacheId) {
          await storage.set(data.key, data.value, data.ttl);
        }
      }
    });
    this._qified.subscribe(deleteEvent, {
      handler: async (message) => {
        const data = message.data;
        if (data.cacheId !== cacheId) {
          await storage.delete(data.key);
        }
      }
    });
  }
  /**
   * Creates or returns a Qified instance from the provided value
   * @param value - Either an existing Qified instance or MessageProvider(s)
   * @returns A Qified instance configured with the provided message provider(s)
   */
  createQified(value) {
    if (value instanceof import_qified.Qified) {
      return value;
    }
    const providers = Array.isArray(value) ? value : [value];
    return new import_qified.Qified({ messageProviders: providers });
  }
  /**
   * Gets the namespace prefix to use for event names
   * @returns The resolved namespace string or undefined
   */
  getNamespace() {
    if (typeof this._namespace === "function") {
      return this._namespace();
    }
    return this._namespace;
  }
  /**
   * Prefixes an event name with the namespace if one is set
   * @param event - The event to prefix
   * @returns The prefixed event name or the original event
   */
  getPrefixedEvent(event) {
    const ns = this.getNamespace();
    return ns ? `${ns}::${event}` : event;
  }
};

// src/index.ts
var import_memory2 = require("@cacheable/memory");
var import_utils2 = require("@cacheable/utils");
var import_keyv2 = require("keyv");
var Cacheable = class extends import_hookified2.Hookified {
  _primary = (0, import_memory.createKeyv)();
  _secondary;
  _nonBlocking = false;
  _ttl;
  _stats = new import_utils.Stats({ enabled: false });
  _namespace;
  _cacheId = Math.random().toString(36).slice(2);
  _sync;
  /**
   * Creates a new cacheable instance
   * @param {CacheableOptions} [options] The options for the cacheable instance
   */
  constructor(options) {
    super();
    if (options?.primary) {
      this.setPrimary(options.primary);
    }
    if (options?.secondary) {
      this.setSecondary(options.secondary);
    }
    if (options?.nonBlocking) {
      this._nonBlocking = options.nonBlocking;
    }
    if (options?.stats) {
      this._stats.enabled = options.stats;
    }
    if (options?.ttl) {
      this.setTtl(options.ttl);
    }
    if (options?.cacheId) {
      this._cacheId = options.cacheId;
    }
    if (options?.namespace) {
      this._namespace = options.namespace;
      this._primary.namespace = this.getNameSpace();
      if (this._secondary) {
        this._secondary.namespace = this.getNameSpace();
      }
    }
    if (options?.sync) {
      this._sync = options.sync instanceof CacheableSync ? options.sync : new CacheableSync({
        ...options.sync,
        namespace: options.namespace
      });
      this._sync.subscribe(this._primary, this._cacheId);
    }
  }
  /**
   * The namespace for the cacheable instance
   * @returns {string | (() => string) | undefined} The namespace for the cacheable instance
   */
  get namespace() {
    return this._namespace;
  }
  /**
   * Sets the namespace for the cacheable instance
   * @param {string | (() => string) | undefined} namespace The namespace for the cacheable instance
   * @returns {void}
   */
  set namespace(namespace) {
    this._namespace = namespace;
    this._primary.namespace = this.getNameSpace();
    if (this._secondary) {
      this._secondary.namespace = this.getNameSpace();
    }
    if (this._sync) {
      this._sync.namespace = namespace;
    }
  }
  /**
   * The statistics for the cacheable instance
   * @returns {CacheableStats} The statistics for the cacheable instance
   */
  get stats() {
    return this._stats;
  }
  /**
   * The primary store for the cacheable instance
   * @returns {Keyv} The primary store for the cacheable instance
   */
  get primary() {
    return this._primary;
  }
  /**
   * Sets the primary store for the cacheable instance
   * @param {Keyv} primary The primary store for the cacheable instance
   */
  set primary(primary) {
    this._primary = primary;
  }
  /**
   * The secondary store for the cacheable instance
   * @returns {Keyv | undefined} The secondary store for the cacheable instance
   */
  get secondary() {
    return this._secondary;
  }
  /**
   * Sets the secondary store for the cacheable instance. If it is set to undefined then the secondary store is disabled.
   * @param {Keyv | undefined} secondary The secondary store for the cacheable instance
   * @returns {void}
   */
  set secondary(secondary) {
    this._secondary = secondary;
  }
  /**
   * Gets whether the secondary store is non-blocking mode. It is set to false by default.
   * If it is set to true then the secondary store will not block the primary store.
   *
   * [Learn more about non-blocking mode](https://cacheable.org/docs/cacheable/#non-blocking-operations).
   *
   * @returns {boolean} Whether the cacheable instance is non-blocking
   */
  get nonBlocking() {
    return this._nonBlocking;
  }
  /**
   * Sets whether the secondary store is non-blocking mode. It is set to false by default.
   * If it is set to true then the secondary store will not block the primary store.
   *
   * [Learn more about non-blocking mode](https://cacheable.org/docs/cacheable/#non-blocking-operations).
   *
   * @param {boolean} nonBlocking Whether the cacheable instance is non-blocking
   * @returns {void}
   */
  set nonBlocking(nonBlocking) {
    this._nonBlocking = nonBlocking;
  }
  /**
   * The time-to-live for the cacheable instance and will be used as the default value.
   * can be a number in milliseconds or a human-readable format such as `1s` for 1 second or `1h` for 1 hour
   * or undefined if there is no time-to-live.
   *
   * [Learn more about time-to-live](https://cacheable.org/docs/cacheable/#shorthand-for-time-to-live-ttl).
   *
   * @returns {number | string | undefined} The time-to-live for the cacheable instance in milliseconds, human-readable format or undefined
   * @example
   * ```typescript
   * const cacheable = new Cacheable({ ttl: '1h' });
   * console.log(cacheable.ttl); // 1h
   * ```
   */
  get ttl() {
    return this._ttl;
  }
  /**
   * Sets the time-to-live for the cacheable instance and will be used as the default value.
   * If you set a number it is miliseconds, if you set a string it is a human-readable
   * format such as `1s` for 1 second or `1h` for 1 hour. Setting undefined means that
   * there is no time-to-live.
   *
   * [Learn more about time-to-live](https://cacheable.org/docs/cacheable/#shorthand-for-time-to-live-ttl).
   *
   * @param {number | string | undefined} ttl The time-to-live for the cacheable instance
   * @example
   * ```typescript
   * const cacheable = new Cacheable();
   * cacheable.ttl = '1h'; // Set the time-to-live to 1 hour
   * ```
   * or setting the time-to-live in milliseconds
   * ```typescript
   * const cacheable = new Cacheable();
   * cacheable.ttl = 3600000; // Set the time-to-live to 1 hour
   * ```
   */
  set ttl(ttl) {
    this.setTtl(ttl);
  }
  /**
   * The cacheId for the cacheable instance. This is primarily used for the wrap function to not have conflicts.
   * If it is not set then it will be a random string that is generated
   * @returns {string} The cacheId for the cacheable instance
   */
  get cacheId() {
    return this._cacheId;
  }
  /**
   * Sets the cacheId for the cacheable instance. This is primarily used for the wrap function to not have conflicts.
   * If it is not set then it will be a random string that is generated
   * @param {string} cacheId The cacheId for the cacheable instance
   */
  set cacheId(cacheId) {
    this._cacheId = cacheId;
  }
  /**
   * Gets the sync instance for the cacheable instance
   * @returns {CacheableSync | undefined} The sync instance for the cacheable instance
   */
  get sync() {
    return this._sync;
  }
  /**
   * Sets the sync instance for the cacheable instance
   * @param {CacheableSync | undefined} sync The sync instance for the cacheable instance
   */
  set sync(sync) {
    th
